Improve throughput throttling for bulks

danielmitterdorfer opened the pull request:
With this commit we allow Rally to throttle throughput not only based on
the number of requests per second but also based on any unit that is
supported by the respective parameter source / runner.

We also introduce a feedback loop between the runner and its scheduler,
allowing the scheduler to react to changes in request granularity (e.g.
adapting to changes in bulk size).

Finally, we also introduce an optional feedback loop between the
scheduler and the parameter source but don't use it in Rally's
implementation.

Existing scheduler implementations are deprecated but retrofitted.

Closes #1069

danielmitterdorfer commented:
Notes to reviewers:

There are two flavors of schedulers and the user documentation covers only simple schedulers. This is intentional as the API is rather new and we should get more experience with it ourselves before exposing it to the entire user base. Therefore, the complete docs are only contained in `scheduler.py` as a module-level docstring.

To test this, you can do the following in rally-tracks:

```diff
diff --git a/geonames/challenges/default.json b/geonames/challenges/default.json
index 0d95152..9fb4680 100644
--- a/geonames/challenges/default.json
+++ b/geonames/challenges/default.json
@@ -247,7 +247,8 @@
         {
           "operation": "index-append",
           "warmup-time-period": 120,
-          "clients": {{bulk_indexing_clients | default(8)}}
+          "clients": {{bulk_indexing_clients | default(8)}},
+          "target-throughput": "90000 docs/s"
         },
         {
           "operation": {
```

Then run a benchmark with

```
esrally --track=geonames --challenge=append-no-conflicts-index-only
```

Here I set the target throughput to "10000 docs/s":

![target-throughput-can-be-reached](https://user-images.githubusercontent.com/1699576/97562945-b1ba1e00-19e2-11eb-8cba-35d15c52c619.png)


Metrics:

|                                                         Metric |         Task |       Value |   Unit |
|---------------------------------------------------------------:|-------------:|------------:|-------:|
|                                                 Min Throughput | index-append |     9984.89 | docs/s |
|                                              Median Throughput | index-append |     10009.4 | docs/s |
|                                                 Max Throughput | index-append |     10198.6 | docs/s |
|                                        50th percentile latency | index-append |     896.864 |     ms |
|                                        90th percentile latency | index-append |     1323.01 |     ms |
|                                        99th percentile latency | index-append |      1564.9 |     ms |
|                                      99.9th percentile latency | index-append |     1916.62 |     ms |
|                                       100th percentile latency | index-append |     1983.08 |     ms |
|                                   50th percentile service time | index-append |     893.018 |     ms |
|                                   90th percentile service time | index-append |     1319.17 |     ms |
|                                   99th percentile service time | index-append |     1560.93 |     ms |
|                                 99.9th percentile service time | index-append |     1915.43 |     ms |
|                                  100th percentile service time | index-append |     1981.22 |     ms |
|                                50th percentile processing time | index-append |     901.862 |     ms |
|                                90th percentile processing time | index-append |     1322.94 |     ms |
|                                99th percentile processing time | index-append |     1563.06 |     ms |
|                              99.9th percentile processing time | index-append |     1917.15 |     ms |
|                               100th percentile processing time | index-append |     1989.42 |     ms |
|                                                     error rate | index-append |           0 |      % |


If I set throughput to "90000 docs/s" (the machine can reach ~ 30000 docs/s):


![target-throughput-cannot-be-reached](https://user-images.githubusercontent.com/1699576/97563059-d7472780-19e2-11eb-9849-d811641c0a03.png)

Metrics:

|                                                         Metric |         Task |       Value |   Unit |
|---------------------------------------------------------------:|-------------:|------------:|-------:|
|                                                 Min Throughput | index-append |     30686.9 | docs/s |
|                                              Median Throughput | index-append |     30905.9 | docs/s |
|                                                 Max Throughput | index-append |     31345.9 | docs/s |
|                                        50th percentile latency | index-append |      162114 |     ms |
|                                        90th percentile latency | index-append |      227943 |     ms |
|                                        99th percentile latency | index-append |      242932 |     ms |
|                                      99.9th percentile latency | index-append |      244835 |     ms |
|                                       100th percentile latency | index-append |      244952 |     ms |
|                                   50th percentile service time | index-append |     1041.73 |     ms |
|                                   90th percentile service time | index-append |     2296.55 |     ms |
|                                   99th percentile service time | index-append |     3565.12 |     ms |
|                                 99.9th percentile service time | index-append |     4096.44 |     ms |
|                                  100th percentile service time | index-append |     4744.81 |     ms |
|                                50th percentile processing time | index-append |     1050.77 |     ms |
|                                90th percentile processing time | index-append |     2311.82 |     ms |
|                                99th percentile processing time | index-append |      3571.3 |     ms |
|                              99.9th percentile processing time | index-append |     4125.55 |     ms |
|                               100th percentile processing time | index-append |     4780.08 |     ms |
|                                                     error rate | index-append |           0 |      % |

gingerwizard commented on esrally/driver/scheduler.py:
```diff
@@ -15,29 +15,114 @@
+def is_simple_scheduler(scheduler_class):
+    """
+    Determines whether a scheduler is a "simple" scheduler, i.e. it doesn't consider feedback from the parameter source.
+    """
```
Is this statement accurate as a "regular" scheduler doesn't have to reference the parameter source just `before_request` and `after_request` - this is more feedback from the runner

danielmitterdorfer replied:
Ah, you're right. I was heavily refactoring this part of the code shortly before creating the PR. That was a leftover. :) I'll change it.

gingerwizard reviewed (commented):
Initial comments mostly to clarify understanding. My main concern is where we call `before_request` and its resulting value. 

I'm also not comfortable with how the parameter source is injected as it doesn't feel like a clean interface but i don't have a better alternative if the intention is for the scheduler to be able to change values in the parameter source. For schedulers which just want to receive values from the parameter source maybe we could always pass through the parameter source params from the runner?

danielmitterdorfer commented on docs/migrate.rst:
```diff
@@ -1,6 +1,63 @@
+        return current + 1
+
+This needs to be reimplemented as follows. We assume that the property ``target-throughput`` is now specified on the respective task instead of hard-coding it in the scheduler. Rally will calculate the correct target throughput and inject it into the scheduler class::
+
```
No. Rally will honor the unit and check that the unit as determined by the parameter source matches the unit specified here. See `UnitAwareScheduler#after_request()`.

gingerwizard replied:
Yes i meant to resolve this once i'd read more of the code.

danielmitterdorfer commented on docs/migrate.rst:
```diff
@@ -1,6 +1,63 @@
+        def next(self, current):
+            return current + self.rate
+
+To stay compatible with Rally 2.0.3, this class needs to be changed as follows::
```
It is intentionally 2.0.3. You're right, also the deprecated API will still work with 2.0.3 but the purpose of this documentation is to show users how to get rid of the deprecated API. I'll reword.

danielmitterdorfer commented on docs/migrate.rst:
```diff
@@ -1,6 +1,63 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Metrics records for bulk request don't contain the ``bulk-size`` property anymore. Please use the ``weight`` property instead and consider the ``unit`` property to interpret the value.
+
```
They could have relied on this property in Kibana visualizations. I also wasn't sure whether it's worth including but I decided that being explicit about the change is better.

danielmitterdorfer commented on esrally/driver/driver.py:
```diff
@@ -1333,12 +1333,14 @@ async def __call__(self, *args, **kwargs):
                         await asyncio.sleep(rest)
                 request_context = self.es["default"].init_request_context()
                 processing_start = time.perf_counter()
+                self.schedule_handle.before_request(processing_start)
```
The intention is that throttling adapts only after a complete interaction, i.e. in the next loop iteration. This is a callback to inform the scheduler that the request is about to start and we only provide the current timestamp for now.

gingerwizard replied:
Discussed with @danielmitterdorfer and although we can't see an immediate need for for `before_request` it means there is some symmetry and might be useful re timing information. Given this API isn't being documented and is subject to change as we use, resolving.

danielmitterdorfer commented on esrally/driver/scheduler.py:
```diff
@@ -15,29 +15,114 @@
+    def before_request(self, now):
+        ...
+
+    def after_request(self, now, weight, unit, meta_data):
```
Regarding the confusing `meta_data` parameter: we could probably rename it. We can also add another method parameter here to add the current request's input parameters from the parameter source.

The intention behind the injected `parameter_source` property is to use it only to close the feedback loop if needed, e.g. to provide feedback to the parameter source to change the bulk size or other parameters due to information gathered in the scheduler. In my opinion this should not be needed usually and hence I'd rather add a new parameter to `after_request`.

gingerwizard replied:
Maybe we hold off adding a parameter for the request's input parameters from the parameter source until its clear we need it. In a lot of cases the user will probably have a custom runner and these parameters could be passed in the runners `meta_data` i.e. a workaround exists. If it becomes clear we need, then we can add.

I would like to rename `meta_data` - maybe something alluding to the fact its from the runner as this won't be obvious.

danielmitterdorfer replied:
Agreed on keeping the parameters as is. I'll rename `meta_data` as discussed.

danielmitterdorfer commented on esrally/driver/runner.py:
```diff
@@ -351,18 +357,16 @@ async def __call__(self, es, params):
         The following meta data are always returned:
 
         * ``index``: name of the affected index. May be `None` if it could not be derived.
-        * ``bulk-size``: bulk size, e.g. 5.000.
```
I feel this is special enough that we should not wait until the next minor. The only use for this property I can think of are Kibana visualizations and they can easily be adapted to use `weight` instead.

danielmitterdorfer commented on esrally/driver/runner.py:
```diff
@@ -332,7 +337,8 @@ async def __call__(self, es, params):
 
         * ``body``: containing all documents for the current bulk request.
-        * ``bulk-size``: the number of documents in this bulk.
+        * ``bulk-size``: An indication of the bulk size denoted in ``unit``.
```
That's tricky because we cannot support `bulk-size` without introducing bottlenecks when the unit is not "docs" (because we need to count the bulk items ourselves which requires deserializing the request or the response which we now avoid at all costs). As I've argued above I feel this is special enough that we can just remove it.

gingerwizard replied:
Discussed with @danielmitterdorfer

danielmitterdorfer commented on esrally/driver/runner.py:
```diff
@@ -494,7 +494,14 @@ async def __call__(self, es, params):
+        if self.first_request:
+            self.first_request = False
+            if "unit" not in params:
+                self.logger.warning("Specify the mandatory parameter [unit] in the bulk parameter source. "
```
Sure, we can discuss this but I agree it's a separate discussion from this PR.

gingerwizard replied:
https://github.com/elastic/rally/issues/1101

danielmitterdorfer commented on esrally/driver/scheduler.py:
```diff
@@ -15,29 +15,114 @@
+
+If the scheduler also needs access to the parameter source, provide a ``parameter_source`` property. Rally injects the
+task's parameter source into this property.
+"""
```
Your understanding is correct: there is one parameter source instance per virtual client. There are a couple of aspects around this feedback loop:

1. This is an advanced feature and should not affect the rest of the API
2. When this feature is required a user will implement a custom scheduler and a custom parameter source so they will be tightly coupled in any case.
3. We want to establish feedback without imposing any API on the parameter source or without assuming anything in particular about its implementation.

gingerwizard replied:
> When this feature is required a user will implement a custom scheduler and a custom parameter source so they will be tightly coupled in any case.
+1.

danielmitterdorfer commented:
@gingerwizard thanks for the feedback. I've addressed your comments in https://github.com/elastic/rally/pull/1100/commits/d3bf0649eeb040b76fa36aed6f9cb7ea5de49e4d and left a couple of inline comments. Happy to discuss more complex aspects over a more high bandwidth medium. :)

gingerwizard reviewed (approved):
LGTM. A few nits @danielmitterdorfer will address.
Main one is the `meta_data` renaming but not prohibitive to being merged as interface is likely to change and is not documented for now.

danielmitterdorfer commented on docs/migrate.rst:
```diff
@@ -1,6 +1,63 @@
+* Scheduler functions should be replaced by scheduler classes
+* The constructor for scheduler classes changes. Instead of receiving ``params``, it receives the entire ``task`` and the precalculated target throughput.
+
+The following scheduler is implemented as a function. The target throughput is hardcoded to one operation per second::
```
Thanks. Addressed in https://github.com/elastic/rally/pull/1100/commits/d3bf0649eeb040b76fa36aed6f9cb7ea5de49e4d.
